### PR TITLE
Bio 2359 yaml config parser

### PR DESCRIPTION
**Pull request by monikaBrandt**
**What problems does this PR solve?**
Added a function that can open, read and parse a yaml.config file for organising files prior to delivery. Can be applied on an anlysed project or a single runfolder. The code will resolve regular expressions, using the pattern given in the configuration file, to find source files to link to a destination file. Based on the input and the resolved regular expressions string substitution will be performed to update/resolve paths given in the configuration file.  

**How has the changes been tested?**
Only unittests.

**Reasons for careful code review**
If any of the boxes below are checked, extra careful code review should be inititated.

  - [ ] This PR contains code that could remove data

### CODE REVIEW

File: tests/unit_tests/services/test_organise_service.py [hunk 1]
```diff
@@ -217,3 +217,71 @@ def test_organise_project_file(self):
...
+        input_key = "runfolder_name"
+        input_value = "200624_A00834_0183_BHMTFYDRXX"
+
+        OrganiseService.get_mock_file_list = mock.MagicMock(return_value=["/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/seqreports/project/AB-1234/200624_A00834_0183_BHMTFYDRXX_AB-1234_multiqc_report.html"])
+
+        expected_output = [("/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/seqreports/project/AB-1234/200624_A00834_0183_BHMTFYDRXX_AB-1234_multiqc_report.html",
+        "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Projects/AB-1234/200624_A00834_0183_BHMTFYDRXX/200624_A00834_0183_BHMTFYDRXX_AB-1234_multiqc_report.html", 
+        {"required": True,
+        "link_type": "softlink",
+        "filter": "(?P<projectid>[\w-]+)/200624_A00834_0183_BHMTFYDRXX_(?P=projectid)_multiqc_report[\w.-]+"})]
```
Comment (Aratz) [inline tests/unit_tests/services/test_organise_service.py:261]:
```suggestion
        expected_output = [
            (
                "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/seqreports/project/AB-1234/200624_A00834_0183_BHMTFYDRXX_AB-1234_multiqc_report.html",
                "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Projects/AB-1234/200624_A00834_0183_BHMTFYDRXX/200624_A00834_0183_BHMTFYDRXX_AB-1234_multiqc_report.html", 
                {
                    "required": True,
                    "link_type": "softlink",
                    "filter": "(?P<projectid>[\w-]+)/200624_A00834_0183_BHMTFYDRXX_(?P=projectid)_multiqc_report[\w.-]+"
                }
            )
        ]
```

File: tests/unit_tests/services/test_organise_service.py [hunk 2]
```diff
@@ -217,3 +217,71 @@ def test_organise_project_file(self):
...
+        input_key = "runfolder_name"
+        input_value = "200624_A00834_0183_BHMTFYDRXX"
+
+        OrganiseService.get_mock_file_list = mock.MagicMock(return_value=["/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Unaligned/AB-1234/Sample_AB-1234-14092/AB-1234-14092_S35_L001_R1_001.fastq.gz"])
+
+        expected_output = [("/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Unaligned/AB-1234/Sample_AB-1234-14092/AB-1234-14092_S35_L001_R1_001.fastq.gz",
+        "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Projects/AB-1234/200624_A00834_0183_BHMTFYDRXX/Sample_AB-1234-14092/AB-1234-14092_S35_L001_R1_001.fastq.gz", 
+        {"required": True,
+        "link_type": "softlink",
+        "filter": "(?P<projectid>[\w-]+)/Sample_(?P<samplename>[\w-]+)/(?P=samplename)_S(?P<samplenumber>\d+)_L(?P<lanes>\d+)_R(?P<read>\d)_001.fastq.gz"})]
```
Comment (Aratz) [inline tests/unit_tests/services/test_organise_service.py:276]:
```suggestion
        expected_output = [
            (
                "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Unaligned/AB-1234/Sample_AB-1234-14092/AB-1234-14092_S35_L001_R1_001.fastq.gz",
                "/proj/ngi2016001/incoming/200624_A00834_0183_BHMTFYDRXX/Projects/AB-1234/200624_A00834_0183_BHMTFYDRXX/Sample_AB-1234-14092/AB-1234-14092_S35_L001_R1_001.fastq.gz", 
                {
                    "required": True,
                    "link_type": "softlink",
                    "filter": "(?P<projectid>[\w-]+)/Sample_(?P<samplename>[\w-]+)/(?P=samplename)_S(?P<samplenumber>\d+)_L(?P<lanes>\d+)_R(?P<read>\d)_001.fastq.gz"
                }
            )
        ]
```

**Review by Aratz — COMMENTED** [review]
Great work! My main concern is with the deadlock issue, I think we should merge the glob functionalities first before you can rework this part.

File: delivery/services/organise_service.py [hunk 3]
```diff
@@ -207,3 +210,94 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+                source_list = OrganiseService.resolve_config_regexp(element["options"]["filter"])
+                for source_file in source_list:
+                    source = ''.join([element["source"].strip('*'),source_file["source_file"]])
+                    destination_file = element["destination"].format_map(Default(source_file))
+                    destination = ''.join([destination_file, source_file["source_file"].split("/")[-1]])
+                    
+                    config_as_list.append((source,destination,element["options"]))
+
+            else:
+                config_as_list.append((element["source"],element["destination"],element["options"]))
```
[the remark this answers is not in the document]
Comment (monikaBrandt) [reply delivery/services/organise_service.py:256]: Yes, this part is a bit tricky and I guess we can solve it in a different way. In this case (i.e. this code will be applied on a runfolder not an organized project) this glob will be resolved when the list of possible source files is created. When a glob is trailing a source path everything under that path should be selected and then filtered with a regexp to get the correct runfolder and project. In the other case, the one you are linking to, the identifier projectid is already specified in the source path and there is only one destination. Therefore the glob can be resolved in the copy/linking operation performed by @b97pla code. But maybe I'm building my code on too many assumptions (or too much knowledge) about the data we are linking.

File: delivery/services/organise_service.py [hunk 4]
```diff
@@ -207,3 +210,94 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
             read_no=sample_file.read_no,
             is_index=sample_file.is_index,
             checksum=sample_file.checksum)
+
+    def parse_yaml_config(config_file_path, input_key, input_value):
```
[the remark this answers is not in the document]
Comment (monikaBrandt) [reply delivery/services/organise_service.py:214]: Do you mean that the default value in the config should be input_key: "runfolder" and during parsing this should be substituted to input_key: <runfolder name> and then we use the key input_key to resolve the paths in the config?

[the remark this answers is not in the document]
Comment (monikaBrandt) [reply delivery/services/organise_service.py:214]: Thank you for good comments and feedback! I was wondering more about where we should set the value of runfolder. Let's say we still have the parameter named input_value and input_key is specified in the config as you show above. Will we then just do
config["input_key"] = input_value and then use {input_key} to do substitutions in the rest of the config? we can absolutely have a look at it tomorrow!

File: delivery/services/organise_service.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Aratz) [reply delivery/services/organise_service.py:256]: I mean I think this function should work for all possible config files, so you should assume that the source path may sometimes resolve to several files, even when filter is not specified. Maybe we can also sort this out irl.

File: delivery/services/organise_service.py [hunk 5]
```diff
@@ -207,3 +210,94 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+            element["source"] = element["source"].format_map(Default(variables_dict))
+            element["destination"] = element["destination"].format_map(Default(variables_dict))
+
+            if "filter" in element["options"].keys(): 
+                element["options"]["filter"] = element["options"]["filter"].format_map(Default(variables_dict))
+                source_list = OrganiseService.resolve_config_regexp(element["options"]["filter"])
+                for source_file in source_list:
+                    source = ''.join([element["source"].strip('*'),source_file["source_file"]])
+                    destination_file = element["destination"].format_map(Default(source_file))
+                    destination = ''.join([destination_file, source_file["source_file"].split("/")[-1]])
```
Comment (b97pla) [inline delivery/services/organise_service.py:251]: I don't think this is what we want in the end but we may need to look into that more closely when we look at real examples.

Also, I'd recommend:
```suggestion
                    destination = os.path.join(destination_file, os.path.basename(source_file["source_file"]))
```

**Review by b97pla — CHANGES_REQUESTED** [review]
Great work! :+1:  

like we discussed, I think the code needs some fixes and I hope we clarified a bit how the different parts of the code should interconnect. I will do some fixing up of my PR

File: delivery/services/organise_service.py [hunk 6]
```diff
@@ -207,3 +291,198 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+        keys in the configuration file and results from resolved 
+        regular expressions (based on patterns from the configuration file). 
+
+        :param config_file_path: Full path to a 
+        configuration file in yaml-format.
+        :param input_value: A runfolder name or a 
+        project name (for analysed projects).
+        :return: A list with one element per link/copy operation to be 
+        performed during organisation.
+         Each element consist of a tuple with 3 values
```
Comment (b97pla) [inline delivery/services/organise_service.py:323]: afaict, each element will be a list and not a tuple (which is perfectly fine)

File: delivery/services/organise_service.py [hunk 7]
```diff
@@ -207,3 +291,198 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+            #Perform string substitutions on variable
+            #  values based on variable keys.
+            variables_dict[key] = OrganiseService.sub_values_with_dict(
+                variables_dict[key], 
+                variables_dict)
+
+        for file_element in config["files_to_organise"]:
+            # for each source-destination pair, 
+            # under "files_to_organise", given in the configuration file
+            config_as_list = config_as_list + OrganiseService.process_file_to_organise(
```
Comment (b97pla) [inline delivery/services/organise_service.py:346]: the `+` operator will add all elements from the second list to the first. use `config_as_list.append()` instead

File: delivery/services/organise_service.py [hunk 8]
```diff
@@ -207,3 +291,198 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+        several source files and destination files if applicable. 
+        Unknowns in the source and destination path will be 
+        resolved by substitutions or by calling resolve_config_regexp().
+        :param file_element: A dictionary with keys: source, destination, options 
+        corresponding to one element of the 
+        "files_to_organise" section in the configuration file.
+        :param variables_dict: A dictionary corresponding to the 
+        section "variables" from the configuration file.
+        :return: A list with all selected files under source. 
+        Each element is a tuple
```
Comment (b97pla) [inline delivery/services/organise_service.py:368]: ... a list

File: delivery/services/organise_service.py [hunk 9]
```diff
@@ -207,3 +291,198 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+        list element, i.e. to each dictionary, the actual match 
+        (the source file) is also added. If there is no
+        regexp pattern given as input the returned dictionary will 
+        only have on key and value.
+        {"source_file": "/path/to/source/file"}
+        """
+        source_file_list = []
+        for file in list_of_source_files:
+            search_re = re.search(regexp_pattern, file)
+
```
Comment (b97pla) [inline delivery/services/organise_service.py:448]: just a suggestion but you could simplify the code below a bit:

```
source_file_dict = {"source_file": file}
if search_re:
  source_file_dict.update(search_re.groupdict())
source_file_list.append(source_file_dict)
```

File: delivery/services/organise_service.py [hunk 10]
```diff
@@ -207,3 +291,198 @@ def organise_sample_file(self, sample_file, organised_sample_path, lanes):
...
+        source_list = OrganiseService.resolve_config_regexp(
+            filter_pattern, 
+            list_of_mock_files)
+        for source_file in source_list:
+            #substitute unknowns in destination path based on 
+            # varibales found in source file
+            destination_file = OrganiseService.sub_values_with_dict(
+                file_element["destination"], 
+                source_file)
+            file_to_organise_list.append(
```
Comment (b97pla) [inline delivery/services/organise_service.py:399]: the way "my" function works, it expects the destination to be a full path, i.e. including filename (if it is a file). In other words, you would need to append `os.path.basename(source_file)` to `destination_file`. Another option is that my function takes care of that, which would be closer to how e.g. `cp` and `ln` works on linux, but I assumed my function would be as dumb as possible.. we can have a quick discussion about this.

**Review by b97pla — APPROVED** [review]
:+1: great work, I think the code is much more readable now!

I had a few comments, mostly just suggestions for further simplification of the code, nothing that affects the functionality (although there were a couple of snags that needs to be fixed).

If you want, I can merge your PR with my code and fix those things directly, or you can update the PR to your preference.

**Comment by b97pla** [comment]
👍 Excellent work @monikaBrandt! I will close this and pull the code into a new PR.

**Pull request closed without merging**